samples: Bluetooth: Broadcast multiple legacy and extended advertising

cvinayak opened the pull request:
Update broadcaster_multiple sample to start multiple advertising sets of type legacy and extended advertising.

How they appear on-air (BabbleSim import), for 4 advertising sets:
<img width="1492" height="138" alt="image" src="https://github.com/user-attachments/assets/ad19c510-4838-438f-a6e3-1681bdc69363" />

cvinayak committed:
Bluetooth: Controller: Fix missing chain pdu time reservation

Fix missing auxiliary chain pdu time reservation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Tune aux offset calculation overhead assertion

Tune the aux offset calculation overhead assertion to use
EVENT_OVERHEAD_START_US; this is the correct maximum
overhead causing ISR latency for start of a radio event that
occurs thereafter.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: Broadcast multiple legacy and extended advertising

Update broadcaster_multiple sample to start multiple
advertising sets of type legacy and extended advertising.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: bsim: Bluetooth: Enable similar Controller Kconfig features

Enable similar Controller Kconfig features between nRF52 and
nRF5340 BabbleSIM LE Audio testing.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on samples/bluetooth/broadcaster_multiple/src/broadcaster_multiple.c:
```diff
@@ -78,20 +82,56 @@ int broadcaster_multiple(void)
 		/* Use advertising set instance index as SID */
 		adv_param.sid = index;
 
+		/* Use Coded PHY */
+		if ((index % 4) == 3) {
+			adv_param.options = BT_LE_ADV_OPT_EXT_ADV;
+			adv_param.options |= BT_LE_ADV_OPT_CODED;
+			ad = ad_long;
+			ad_size = ARRAY_SIZE(ad_long);
+		/* Use 1M auxiliary PDU */
+		} else if ((index % 4) == 2) {
+			adv_param.options = BT_LE_ADV_OPT_EXT_ADV;
+			adv_param.options |= BT_LE_ADV_OPT_NO_2M;
+			ad = ad_long;
+			ad_size = ARRAY_SIZE(ad_long);
+		/* Use 2M auxiliary PDU */
+		} else if ((index % 4) == 1) {
+			adv_param.options = BT_LE_ADV_OPT_EXT_ADV;
+			ad = ad_long;
+			ad_size = ARRAY_SIZE(ad_long);
+		/* Use 1M legacy PDU */
+		} else {
+			adv_param.options = BT_LE_ADV_OPT_NONE;
+			ad = ad_short;
+			ad_size = ARRAY_SIZE(ad_short);
+		}
```
Consider moving this into a `select` function to simplify the loop. 

Also consider storing this in an array like

```c
struct {
    uint32_t options;
    const struct bt_data *ad;
    size_t ad_size
} adv_params[4];
```

So you can do things like `(index % sizeof(adv_params) == 2U)`. Alternatively you can also make it a circular linked list and just get the next entry for each iteration.

Thalley commented on samples/bluetooth/broadcaster_multiple/src/broadcaster_multiple.c:
```diff
@@ -78,20 +82,56 @@ int broadcaster_multiple(void)
-			printk("Failed to create advertising set %d (err %d)\n",
-			       index, err);
+			/* Failed creating Coded PHY advertising set? */
+			if (adv_param.options & BT_LE_ADV_OPT_CODED) {
```
```suggestion
			if ((adv_param.options & BT_LE_ADV_OPT_CODED) != 0) {
```

Thalley commented on samples/bluetooth/hci_ipc/nrf5340_cpunet_iso-bt_ll_sw_split.conf:
Should the changes to this file also be applied to the other `bt_ll_sw_split` files in `hci_ipc`?

Thalley reviewed (commented):
I'll leave the same comment here as I usually do when you modify the samples for testing purposes: IMO samples should not be used for this purpose, and should be kept as simple as possible. I don't think the changes to the sample here will help users, and it seems to only be done to help test some features, which should done in a separate test application, instead of (ab)using samples for test purposes

cvinayak commented on samples/bluetooth/broadcaster_multiple/src/broadcaster_multiple.c:
```diff
@@ -54,50 +54,105 @@
+			const struct bt_data *ad;
+			size_t ad_size;
+		} param_config[] = {
+			/* Use 1M legacy PDU */
+			{.options = BT_LE_ADV_OPT_NONE,
+			 .ad = ad_short,
+			 .ad_size = ARRAY_SIZE(ad_short),
+			},
+			/* Use 2M auxiliary PDU */
+			{.options = BT_LE_ADV_OPT_EXT_ADV,
```
Thanks, will fix this indentation.

cvinayak reviewed (commented):
This PR is not super critical or is not fixing a bug, will take it at my spare time... will put into draft.

cvinayak committed:
samples: Bluetooth: observer: Move Zephyr Controller Kconfigs

Move Zephyr Controller Kconfigs to overlay file.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented on samples/bluetooth/broadcaster_multiple/src/broadcaster_multiple.c:
```diff
@@ -54,44 +54,95 @@
+			const struct bt_data *ad;
+			size_t ad_size;
+		} param_config[] = {
+		{ /* Use 1M legacy PDU */
+			.options = BT_LE_ADV_OPT_NONE,
+			.ad = ad_short,
+			 .ad_size = ARRAY_SIZE(ad_short),
+		},
+		{ /* Use 2M auxiliary PDU */
+			.options = BT_LE_ADV_OPT_EXT_ADV,
+			.ad = ad_long,
+			.ad_size = ARRAY_SIZE(ad_long),
+		},
+		{ /* Use 1M auxiliary PDU */
+			.options = BT_LE_ADV_OPT_EXT_ADV | BT_LE_ADV_OPT_NO_2M,
+			.ad = ad_long,
+			.ad_size = ARRAY_SIZE(ad_long),
+		},
+		{ /* Use Coded PHY */
+			.options = BT_LE_ADV_OPT_EXT_ADV | BT_LE_ADV_OPT_CODED,
+			.ad = ad_long,
+			.ad_size = ARRAY_SIZE(ad_long),
+		}};
```
It feels redundant, and hence adopted the way `case` in a `switch` block is indented as per kernel coding style.

Should I have something like this otherwise?
```
		} param_config[] = {{
			/* Use 1M legacy PDU */
			.options = BT_LE_ADV_OPT_NONE,
			.ad = ad_short,
			.ad_size = ARRAY_SIZE(ad_short),
		}, {
```

cvinayak commented on samples/bluetooth/broadcaster_multiple/prj.conf:
```diff
@@ -1,23 +1,10 @@
-# CONFIG_BT_CTLR_ADVANCED_FEATURES=y
-# CONFIG_BT_CTLR_ADV_DATA_BUF_MAX=2
+# Increased to 4 gives one each of legacy, 2M, 1M and Coded PHY advertising sets
+CONFIG_BT_EXT_ADV_MAX_ADV_SET=2
```
No not all vendors will support 4 as default.

Thalley replied:
I think the issue is that the comment states "Increased to 4" but the option assigns 2

cvinayak replied:
@Thalley Updated the comment.

@jhedberg could you confirm the configuration is not a problem for your implementations?

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -3208,7 +3205,7 @@ static void mfy_aux_offset_get(void *param)
 	ticks_to_expire = 0U;
 	ticks_current = adv->ticks_at_expire;
-	retry = 1U; /* Assert on first ticks_current change */
+	retry = 4U; /* Assert on 2 other adv set ticks_current change, and 2 scanners expiring */
```
Of the 4 advertising sets tested by the sample, 3 advertising sets being extended advertising that calculate the aux_offset can have the other two preempt this code path. Also, 1M PHY and Coded PHY scan window can preempt. Hence, worst case 2 advertising set's primary radio event and 2 scan window's radio event.

May be I should change this to the adv set count + 2 scan windows...

cvinayak commented on tests/bsim/bluetooth/host/adv/chain/src/main.c:
```diff
@@ -146,7 +153,7 @@ static void test_scan_main(void)
 
 	/* Let the recv callback verify the reports */
-	k_sleep(K_SECONDS(10));
+	k_sleep(K_SECONDS(30));
```
the 3 auxiliary PDUs of the sets are very closely packed... and interleaving of the primary to auxiliary needs to fall in an ideal position :-)  to get the report !

cvinayak committed:
samples: Bluetooth: Move harness and tags to common section

Move harness and tags to common section for observer and
broadcaster_multiple samples.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: bsim: Bluetooth: Do not use the BT_ prefix

Do not use the BT_ prefix for definitions not provided by
the Bluetooth subsystem.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented on samples/bluetooth/broadcaster_multiple/overlay-bt_ll_sw_split.conf:
Rename this file to `bt_ll_sw_split.conf`

cc @nordicjm

cvinayak commented on samples/bluetooth/observer/overlay-bt_ll_sw_split.conf:
Rename this file to `bt_ll_sw_split.conf`

cc @nordicjm

cvinayak commented on samples/bluetooth/hci_uart_3wire/overlay-all-bt_ll_sw_split.conf:
Rename this file to `bt_ll_sw_split-all.conf`

cc @nordicjm

cvinayak commented on tests/bsim/bluetooth/audio/overlay-bt_ll_sw_split.conf:
Rename this file to `bt_ll_sw_split.conf`

cc @nordicjm

Thalley commented on tests/bsim/bluetooth/host/adv/chain/src/main.c:
```diff
@@ -76,20 +81,21 @@ static bool data_cb(struct bt_data *data, void *user_data)
 		      struct net_buf_simple *buf)
 {
-	static uint8_t sid[CONFIG_BT_EXT_ADV_MAX_ADV_SET];
+	static uint8_t sid[MAX_EXT_ADV_SET];
```
If `#define MAX_EXT_ADV_SET (CONFIG_BT_EXT_ADV_MAX_ADV_SET - 1)`, then won't we risk `sid[0]` if this is built with `CONFIG_BT_EXT_ADV_MAX_ADV_SET=1`?

If we never want this test to support that config value, suggest to add a BUILD_ASSERT with an explanation

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -3047,9 +3047,17 @@ static uint32_t aux_time_get(const struct ll_adv_aux_set *aux,
-		 *        if any.
-		 */
+#if defined(CONFIG_BT_CTLR_ADV_AUX_PDU_LINK)
+		struct pdu_adv *pdu_chain;
```
done

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -3323,11 +3336,17 @@ static void mfy_aux_offset_get(void *param)
+	ticks_to_start = HAL_TICKER_US_TO_TICKS(EVENT_OVERHEAD_XTAL_US);
+	ticks_to_start += HAL_TICKER_US_TO_TICKS(EVENT_OVERHEAD_START_US);
+	LL_ASSERT_MSG((ticks_elapsed <= ticks_to_start), "%s: elapsed = %u (%u) us.",
+		      __func__, HAL_TICKER_TICKS_TO_US(ticks_elapsed),
```
Isn't the line number printed as part of the assert? if so, adding the function name seems superfluous

Thalley replied:
I see. I guess we could consider handling this in the same way we do for LOG statements, where the function name is optionally printed, but that's of course more generic that just adding it a single place like here

Thalley reviewed (dismissed):
Minor comments, but no blockers

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -3047,9 +3047,17 @@ static uint32_t aux_time_get(const struct ll_adv_aux_set *aux,
-		/* FIXME: Calculate additional time reservations for chain PDUs,
-		 *        if any.
-		 */
+#if defined(CONFIG_BT_CTLR_ADV_AUX_PDU_LINK)
```
`CONFIG_BT_CTLR_ADV_AUX_PDU_BACK2BACK_AFS` in `EVENT_B2B_MAFS_US` is not defined.

cvinayak replied:
`CONFIG_BT_CTLR_ADV_AUX_PDU_BACK2BACK_AFS` that defines `EVENT_B2B_MAFS_US`  is the auxiliary frame spacing value defined by Kconfig and depends on `CONFIG_BT_CTLR_ADV_AUX_PDU_LINK`. To use `IS_ENABLED` for the block, `CONFIG_BT_CTLR_ADV_AUX_PDU_BACK2BACK_AFS` will need to have its `depends on` removed.

Thalley replied:
Ah, so the definition of `EVENT_B2B_MAFS_US` has a hard dependency on `CONFIG_BT_CTLR_ADV_AUX_PDU_BACK2BACK_AFS` due to its definition. That makes sense. 

I do also see that `lll_adv_pdu_linked_next_get` is guarded by `#if defined(CONFIG_BT_CTLR_ADV_PDU_LINK)` (which is a violation of https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#rule-a-1-conditional-compilation but I think the rule came after the declaration in this case), so we do need to use `#if defined(CONFIG_BT_CTLR_ADV_AUX_PDU_LINK)` due to this :)

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_vendor.h:
```diff
@@ -16,26 +16,26 @@
 #endif /* !CONFIG_BT_CTLR_PHY_CODED */
 #else /* !CONFIG_BT_OBSERVER */
-/* Active connection in peripheral role with legacy scanning on 1M.
+/* Simultaneous 3 extended advertising sets calculating aux offsets.
```
@cvinayak This isn't done - The text still says `Simultaneous 3`

Thalley reviewed (approved):
No blockers
